### PR TITLE
Implemented pre-connection JITMs

### PR DESCRIPTION
**Pull request by pidgeonman**
This is a Jetpack plugin-only (pre-connection) port of the WPCOM JITM engine, aimed to present messages to users mostly to advertise Jetpack's features and encourage them to connect.

#### Changes proposed in this Pull Request:
* Implement a Jetpack-only (pre-connection) port of the WPCOM JITM engine to display JITMs without "phoning home".

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This is a new feature, mostly discussed in pbtFPC-nZ-p2 and related posts.

#### Testing instructions:
1) Get a brand new Jetpack environment, activate the Jetpack plugin but DO NOT connect it.
2) Currently there are 3 JITMs implemented **AS EXAMPLES** as specified in pbtFPC-nZ-p2, so **their text could change in the future before the release, or there could be more added**. They should be displayed at edit.php, upload.php, widgets.php like this:

preconnection

3) In addition to the JITMs rendering, click the blue "Setup Jetpack" buttons and check it takes you to the connection screen.
 
For a complete test:
1) Apply D42181-code.
2) Connect Jetpack.
3) You should now be able to see all the post-connection JITMs features already available and NOT the pre-connection ones.

#### Proposed changelog entry for your changes:
* Implemented pre-connection JITMs.

### CODE REVIEW

**Review by jeherve — CHANGES_REQUESTED** [review]
I am running into some errors when post-connection JITMs right now; they are not displayed on my site.

```
[21-Apr-2020 07:32:56 UTC] PHP Notice:  Trying to get property 'hook' of non-object in /var/www/html/wp-content/plugins/jetpack/packages/jitm/src/class-post-connection-jitm.php on line 213
```

```
[21-Apr-2020 07:39:27 UTC] PHP Notice:  Undefined property: stdClass::$cta in /var/www/html/wp-content/plugins/jetpack/packages/jitm/src/class-post-connection-jitm.php on line 213
```

```
Uncaught TypeError: Cannot read property 'message' of undefined
    at Object.default (jetpack-jitm.js?ver=1.0:6)
```

*******

The message on the Widgets screen says:

> great additional widgets like business hours and maps

We do not have a Business Hours or a Map widget so we may have to update this.

*******

Since we have quite a few new methods, it would be really useful to add tests to cover all of this.

**Comment by folletto** [comment]
> port of the WPCOM JITM engine

Do we really need to port the entire engine and 2.3k LOC for 3 banners?
Maybe is there a goal for that that is not self-evident from this specific PR?

File: packages/jitm/src/class-jitm.php [hunk 1]
```diff
@@ -31,70 +33,33 @@ class JITM {
...
-		 *
-		 * @since 3.7.0
-		 * @since 5.4.0 Correct docblock to reflect default arg value
-		 *
-		 * @param bool false Whether to show just in time messages.
-		 */
-		if ( ! apply_filters( 'jetpack_just_in_time_msgs', false ) ) {
-			return false;
+	public static function get_instance() {
+		if ( apply_filters( 'jetpack_just_in_time_msgs', false ) ) {
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-jitm.php]: Fixed.

File: packages/jitm/src/class-cache.php [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+			}
+		);
+	}
+
+	/**
+	 * Get the user's country
+	 *
+	 * The returned country code would be one of the supported countries listed in
+	 * https://opengrok.a8c.com/source/xref/trunk/wp-content/lib/communication/phone-utils.php
+	 *
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-cache.php]: Removed.

File: packages/jitm/src/class-cache.php [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+			'user_country',
+			$user_id,
+			static function () use ( $user_id ) {
+				$country = get_user_attribute( $user_id, 'geo_country_signup' ) !== null ?
+					get_user_attribute( $user_id, 'geo_country_signup' ) :
+					get_user_attribute( $user_id, 'twostep-phone-country' );
+				return isset( $country ) ? $country : '';
+			}
+		);
+	}
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-cache.php]: Removed.

File: packages/jitm/src/class-cache.php [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+			function () use ( $id ) {
+				$user = $id;
+				if ( 0 === $user ) {
+					return false;
+				}
+				$mobile_last_seen = get_user_attribute( $user, 'mobile_app_last_seen' );
+				return ! empty( $mobile_last_seen );
+			}
+		);
+	}
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-cache.php]: Removed.

File: packages/jitm/src/class-cache.php [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+	 */
+	public function get_site_slug() {
+		return $this->get_or_set(
+			'site',
+			'site_slug',
+			static function () {
+				return \Jetpack::build_raw_urls( home_url() );
+			}
+		);
+	}
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-cache.php]: Removed.

File: packages/jitm/src/class-cache.php [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+					foreach ( $module_list as $module ) {
+						$modules[] = $module;
+					}
+				}
+
+				// Clean duplicates.
+				return (array) array_unique( $modules );
+			}
+		);
+	}
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-cache.php]: Removed.

File: packages/jitm/src/class-message.php [hunk 7]
```diff
@@ -0,0 +1,1562 @@
...
+	private function score_multisite( $path, $external_user_id, $query, $score ) {
+		if ( is_null( $this->multisite_check ) ) {
+			return $score;
+		}
+
+		$is_multisite = $this->cache->get_or_set(
+			'callable',
+			'is_multisite',
+			function () {
+				return (bool) $this->cache->get_replica_store()->get_callable( 'is_multi_site' );
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-message.php]: Removed.

File: packages/jitm/src/class-message.php [hunk 8]
```diff
@@ -0,0 +1,1562 @@
...
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_partner_hosted( $path, $external_user_id, $query, $score ) {
+		if ( is_null( $this->hosted_with_partner ) ) {
+			return $score;
+		}
+
+		require_lib( 'jetpack-start' );
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-message.php]: Removed.

File: packages/jitm/src/class-message.php [hunk 9]
```diff
@@ -0,0 +1,1562 @@
...
+		}
+
+		foreach ( $this->stickers as $sticker ) {
+			if ( has_blog_sticker( $sticker ) ) {
+				return 1;
+			}
+		}
+
+		return false;
+	}
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-message.php]: Removed.

File: packages/jitm/src/class-message.php [hunk 10]
```diff
@@ -0,0 +1,1562 @@
...
+		}
+
+		require_lib( 'tos/updates' );
+		$current_user_id = $this->cache->get_current_user( $external_user_id );
+		if ( \A8C\TOS\display_accept_prompt( $current_user_id ) ) {
+			return 1;
+		}
+
+		return false;
+	}
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-message.php]: Removed.

File: packages/jitm/src/class-pre-connection-jitm.php [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+				'jetpack_page_akismet-key-config',
+				'admin_page_jetpack_modules',
+			),
+			true
+		) ) {
+			add_action( 'admin_enqueue_scripts', array( $this, 'jitm_enqueue_files' ) );
+			add_action( 'admin_notices', array( $this, 'ajax_message' ) );
+			add_action( 'edit_form_top', array( $this, 'ajax_message' ) );
+		}
+	}
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-pre-connection-jitm.php]: Refactored.

File: packages/jitm/src/class-pre-connection-jitm.php [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+				unset( $envelopes[ $idx ] );
+				continue;
+			}
+
+			$this->tracking->record_user_event(
+				'jitm_view_client',
+				array(
+					'jitm_id' => $envelope->id,
+				)
+			);
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-pre-connection-jitm.php]: Removed.

File: packages/jitm/src/class-pre-connection-jitm.php [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+
+			// Get affiliate code and add it to the array of URL parameters.
+			$aff = Partner::init()->get_partner_code( Partner::AFFILIATE_CODE );
+			if ( '' !== $aff ) {
+				$url_params['aff'] = $aff;
+			}
+
+			$envelope->url = add_query_arg( $url_params, 'https://jetpack.com/redirect/' );
+
+			$envelope->jitm_stats_url = \Jetpack::build_stats_url( array( 'x_jetpack-jitm' => $envelope->id ) );
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-pre-connection-jitm.php]: Removed.

File: packages/jitm/src/class-pre-connection-jitm.php [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+					break;
+				default:
+					$envelope->content['icon'] = '';
+					break;
+			}
+
+			$jetpack = \Jetpack::init();
+			$jetpack->stat( 'jitm', $envelope->id . '-viewed-' . JETPACK__VERSION );
+			$jetpack->do_stats( 'server_side' );
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-pre-connection-jitm.php]: Removed.

File: packages/jitm/src/class-pre-connection-jitm.php [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+
+	/**
+	 * Is the current page a block editor page?
+	 *
+	 * @since 8.0.0
+	 */
+	public function is_gutenberg_page() {
+		$current_screen = get_current_screen();
+		return ( method_exists( $current_screen, 'is_block_editor' ) && $current_screen->is_block_editor() );
+	}
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-pre-connection-jitm.php]: Removed.

File: packages/jitm/src/class-pre-connection-jitm.php [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+			true
+		) ) {
+			add_action( 'admin_enqueue_scripts', array( $this, 'jitm_enqueue_files' ) );
+			add_action( 'admin_notices', array( $this, 'ajax_message' ) );
+			add_action( 'edit_form_top', array( $this, 'ajax_message' ) );
+		}
+	}
+
+	/**
+	 * Asks the wpcom API for the current message to display keyed on query string and message path
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-pre-connection-jitm.php]: Reworded.

File: packages/jitm/src/class-jitm.php [hunk 17]
```diff
@@ -31,70 +33,33 @@ class JITM {
...
+			$jitm = new Post_Connection_JITM();
+		} else {
+			$jitm = new Pre_Connection_JITM();
 		}
```
[the remark this answers is not in the document]
Comment (pidgeonman) [reply packages/jitm/src/class-jitm.php]: Refactored using Jetpack::is_active().

**Comment by pidgeonman** [comment]
> I am running into some errors when post-connection JITMs right now; they are not displayed on my site.
> 

As stated on the testing instructions, you should apply D42181 first if you want to test the post-connection JITMs too.

> The message on the Widgets screen says:
> 
> > great additional widgets like business hours and maps
> 
> We do not have a Business Hours or a Map widget so we may have to update this.
> 
> Since we have quite a few new methods, it would be really useful to add tests to cover all of this.

These 3 messages are not definitive, they are just placeholders. Its contents may change and others can be added before and after the next release.

**Comment by pidgeonman** [comment]
> > port of the WPCOM JITM engine
> 
> Do we really need to port the entire engine and 2.3k LOC for 3 banners?
> Maybe is there a goal for that that is not self-evident from this specific PR?

@folletto those 3 messages are just placeholders. Their contents will change and more messages can be added before and after the next Jetpack release, even for other hosts. The goal was to reuse code we already have in order to maintain the same JS code at the client side and thus keep the same interface and way to code the JITMs. I don't think coding something adhoc just for the pre-connection JITMs will be wise, specially now I've refactored and simplified the first iteration.

**Comment by pidgeonman** [comment]
@jeherve Thanks for the thorough review! Some comments inline:

> I am running into some errors when post-connection JITMs right now; they are not displayed on my site.
> 

As stated on the testing instructions, you should apply D42181 first if you want to test the post-connection JITMs too.

> The message on the Widgets screen says:
> 
> > great additional widgets like business hours and maps
> 
> We do not have a Business Hours or a Map widget so we may have to update this.
> 
> Since we have quite a few new methods, it would be really useful to add tests to cover all of this.

These 3 messages are not definitive, they are just placeholders. Its contents may change and others can be added before and after the next release.

I have just refactored and simplified the first iteration. This is now unfinished code (not ready for release) until we figure out a proper way to detect hosts.

**Comment by jeherve** [comment]
> As stated on the testing instructions, you should apply D42181-code first if you want to test the post-connection JITMs too.

Oh, I had missed that part, sorry!

> These 3 messages are not definitive, they are just placeholders. Its contents may change and others can be added before and after the next release.

We try to keep `master` as stable as possible and in a "ready to ship" state at all times. Since the messages would be displayed to all new Jetpack site owners and aren't hidden behind a flag or anything that way, I would prefer if the wording on those was correct before we merge them into `master`.

**Comment by folletto** [comment]
> I don't think coding something adhoc just for the pre-connection JITMs will be wise, specially now I've refactored and simplified the first iteration.

I understand, but let's not be hung on a sunken cost. The fact work was put in is never a good reason to move forward with something. I get it can be frustrating, but that's not a reason. 

You also didn't answer the question, if anything, you confirmed the whole JITM logic isn't necessary here when you say we need still to introduce a different logic for hosts. What would also be the cost in maintaining this piece of the JITM engine if the other gets updated?

In short: these aren't JITMs. Why is the JITM engine needed?

File: packages/jitm/src/class-engine.php [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+	/**
+	 * Returns the pre-connection JITMs default rules
+	 *
+	 * @return array Pre-connection JITMs default rules.
+	 */
+	private function preconnection_default_rules() {
+
+		return array(
+			( new Message( 'jpsetup-posts', 'pre-connect' ) )
+				->user_is( 'administrator' )
```
Comment (ebinnion) [inline packages/jitm/src/class-engine.php:48]: Since all of JITMs are meant to be shown to administrators, and since we want an administrator to connect the site, we can likely simplify and get rid of this.

Likewise, since all of the following filters are the same, I'd suggest just getting rid of those:  `->with_icon()`, `->priority( 100 )`, `->open_cta_in_same_window()`, `->is_dismissible( true )`, `->priority( 1000 )`, and `->is_hosted_with_partner( 'bluehost' )`.

File: packages/jitm/src/class-engine.php [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+	 * Gets the top messages
+	 *
+	 * @param string $message_path Message path.
+	 * @param string $user_level User level.
+	 * @param string $query Message query.
+	 * @param bool   $mobile_browser Uses mobile browser.
+	 *
+	 * @return array Rendered messages.
+	 */
+	public function get_top_messages( $message_path, $user_level = '', $query = '', $mobile_browser = false ) {
```
Comment (ebinnion) [inline packages/jitm/src/class-engine.php:120]: Since the JITMs that we're showing all have the same priority but are for different places in wp-admin, do we need to add this logic that is attempts to get the top messages?

File: packages/jitm/src/class-message.php [hunk 20]
```diff
@@ -0,0 +1,1168 @@
...
+	/**
+	 * Score user roles
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_user_roles( $path, $query, $score ) {
```
Comment (ebinnion) [inline packages/jitm/src/class-message.php:409]: I don't think we need this if we're only showing pre-connection JITMs to admins. Likewise with other scoring methods.

File: packages/jitm/src/class-message.php [hunk 21]
```diff
@@ -0,0 +1,1168 @@
...
+	/**
+	 * Score active widgets
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_active_widgets( $path, $query, $score ) {
```
Comment (ebinnion) [inline packages/jitm/src/class-message.php:565]: Are we using any rules that rely on widgets?

File: packages/jitm/src/class-message.php [hunk 22]
```diff
@@ -0,0 +1,1168 @@
...
+	/**
+	 * Score plugins
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_plugins( $path, $query, $score ) {
```
Comment (ebinnion) [inline packages/jitm/src/class-message.php:464]: Are we using any rules that rely on plugins?

File: packages/jitm/src/class-message.php [hunk 23]
```diff
@@ -0,0 +1,1168 @@
...
+	/**
+	 * Score user theme
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_user_theme( $path, $query, $score ) {
```
Comment (ebinnion) [inline packages/jitm/src/class-message.php:433]: Are we using any rules that rely on the theme?

File: packages/jitm/src/class-message.php [hunk 24]
```diff
@@ -0,0 +1,1168 @@
...
+	/**
+	 * Score user locale
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_user_locale( $path, $query, $score ) {
```
Comment (ebinnion) [inline packages/jitm/src/class-message.php:390]: Are we using rules that rely on the user locale?

File: packages/jitm/src/class-message.php [hunk 25]
```diff
@@ -0,0 +1,1168 @@
...
+	/**
+	 * Score option matches
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_option_matches( $path, $query, $score ) {
```
Comment (ebinnion) [inline packages/jitm/src/class-message.php:307]: Are we using rules that rely on options?

**Review by ebinnion — CHANGES_REQUESTED** [review]
Looking at this PR, there is still a lot of unused code that we have pulled over from WPCOM in the form of filters and scoring logic. I believe the overall goal here was to pull over the same logic that we use for generating JITMs on WP.com in an effort to make creating JITMs simpler and quicker in the future. But, that seems unnecessary to me.

The pre-connection JITMs are much simpler here in that we're really only concerned with a few cases:

- Is the site connected or not?
- Is the user an admin?
- Is the JITM customized to the URL path that the user is currently on?
- Is the JITM being shown for the right host?

With the above in mind, we want the user to do the same exact same thing, which is connect their site. With that scope, I believe that the logic can be greatly simplified for showing pre-connection JITMs.

Personally, I'd probably take the following approach:

- Create a PR that hard-codes a pre-connection JITM with something like "Hello World"
  - In that PR, we'd want to pull from the current logic used to display JITMs so that the design is consistent
  - In that same PR, only show the JITM if a filter returns true. We can then use this as a feature flag, allowing hosting partners to turn on/off pre-connection JITMs as they'd like.
  - Some benefits there are that we 1) don't have to worry about hosting provider detection and 2) hosting providers can turn off pre-connection JITMs if there are some unforeseen issues
- Create another PR that handles dismissing a pre-connection JITM
- Create another PR that correctly shows the three or so JITMs based on the URL path that the user is at

This approach is obviously not as robust as creating a whole JITM engine. But, it's light and quick as a result. And let's keep in mind that this pre-connection JITM functionality is really just a test.

**Comment by ebinnion** [comment]
@pidgeonman asked me to take another look at this before he closes the PR.

The biggest issue that I see with the PR in its current state is that it's still loading the Tracks functionality. You can see this for yourself if you view source when one of the JITMs is displayed. In my test site, I'm seeing:

```
<script>
--
  | var jpTracksAJAX = {"ajaxurl":"http:\/\/...\/wp-admin\/admin-ajax.php","jpTracksAJAX_nonce":"..."};
  | </script>
  | <script src='http://.../wp-content/plugins/jetpack/_inc/lib/tracks/tracks-ajax.js?ver=8.4-alpha'></script>
```

This is happening because the pre-connection JITM class extends the post-connection JITM class, which instantiates Tracks in its constructor. One way we could solve that is by creating a base JITM class, that didn't instantiate Tracks. Then, create separate post-connection and pre-connection JITM classes that extend the base. Then, only instantiate Tracks and make API calls in the pre-connection class. 

At less than 600 lines of code now,  this PR has trimmed a huge 2,000 lines of code. That's great. But, I still think that there's some room to simplify. Since these pre-connection JITMs only care about the message path, the JITM engine and message classes seem a bit much to me. I wonder if we could just use something like an associate array, keyed by the message path, in the pre-connection JITM class?

**Comment by pidgeonman** [comment]
Many thanks @ebinnion for taking time to review that last commit. Much appreciated.

I'd like to point out one thing:

> This is happening because the pre-connection JITM class extends the post-connection JITM class

That is not quite right. The pre-connection and post-connection classes both extend the JITM class:

https://github.com/Automattic/jetpack/blob/afacc22062397046d958e5b8e060df1e77ed80f0/packages/jitm/src/class-pre-connection-jitm.php#L16

So in order to avoid enabling the tracking on the pre-connection JITMs, I should move that constructor from the JITM class to the post-connection class which should do the tracks enabling only for it.

I will keep that in mind for the next iteration in a different PR.

Thanks again.
